src/control/tag.rs

```diff
@@ -0,0 +1,81 @@
+use core::{fmt, mem};
+
+/// Single tag in a control group.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub(crate) struct Tag(pub(super) u8);
+impl Tag {
+    /// Control tag value for an empty bucket.
+    pub(crate) const EMPTY: Tag = Tag(0b1111_1111);
+
+    /// Control tag value for a deleted bucket.
+    pub(crate) const DELETED: Tag = Tag(0b1000_0000);
+
+    /// Checks whether a control tag represents a full bucket (top bit is clear).
+    #[inline]
+    pub(crate) const fn is_full(self) -> bool {
+        self.0 & 0x80 == 0
+    }
+
+    /// Checks whether a control tag represents a special value (top bit is set).
+    #[inline]
+    pub(crate) const fn is_special(self) -> bool {
+        self.0 & 0x80 != 0
+    }
+
+    /// Checks whether a special control value is EMPTY (just check 1 bit).
+    #[inline]
+    pub(crate) const fn special_is_empty(self) -> bool {
+        debug_assert!(self.is_special());
+        self.0 & 0x01 != 0
+    }
+
+    /// Creates a control tag representing a full bucket with the given hash.
+    #[inline]
+    #[allow(clippy::cast_possible_truncation)]
+    pub(crate) const fn full(hash: u64) -> Tag {
+        // Constant for function that grabs the top 7 bits of the hash.
+        const MIN_HASH_LEN: usize = if mem::size_of::<usize>() < mem::size_of::<u64>() {
+            mem::size_of::<usize>()
+        } else {
+            mem::size_of::<u64>()
+        };
+
+        // Grab the top 7 bits of the hash. While the hash is normally a full 64-bit
+        // value, some hash functions (such as FxHash) produce a usize result
+        // instead, which means that the top 32 bits are 0 on 32-bit platforms.
+        // So we use MIN_HASH_LEN constant to handle this.
+        let top7 = hash >> (MIN_HASH_LEN * 8 - 7);
+        Tag((top7 & 0x7f) as u8) // truncation
+    }
+}
+impl fmt::Debug for Tag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.is_special() {
+            if self.special_is_empty() {
+                f.pad("EMPTY")
+            } else {
+                f.pad("DELETED")
+            }
+        } else {
+            f.debug_tuple("full").field(&(self.0 & 0x7F)).finish()
+        }
+    }
+}
```

Comment:
I added this to replace the built-in debug to help me debug the code I'm going to add in a future PR, so it's a bit easier to read. I figure that if we care about the additional compile time added by this one debug impl so much we can `cfg(debug_assertions)` gate it later.

Reply:
This isn't a problem for compilation time since it's not a generic function.